src/pykiso/lib/connectors/cc_process.py

```diff
@@ -0,0 +1,265 @@
+        :param name: The name of the stream
+
+        :return: The thread object
+        """
+        thread = threading.Thread(
+            name=f"cc_process_{name}", target=self._read_thread, args=(stream, name)
+        )
+        thread.start()
+        return thread
+
+    def _read_thread(self, stream: IO, name: str) -> None:
+        """Thread for reading data from stdout or stderr
+
+        :param stream: The stream to read from
+        :param name: The name of the stream
+        """
+        try:
+            while True:
+                if self.text:
+                    data = stream.readline()
+                else:
+                    data = stream.read(1)
+                if len(data) == 0:
+                    break
+                # print(data)
+                self.queue_in.put(ProcessMessage(name, data))
+        finally:
+            with self.lock:
+                self.ready += 1
+                if self.ready == int(self.pipe_stdout) + int(self.pipe_stderr):
+                    # None marks the termination of all read threads
+                    self.queue_in.put(None)
+
+    def _cc_close(self) -> None:
+        self._cleanup()
+
+    def _cc_send(self, msg: MessageType, raw: bool = False, **kwargs) -> None:
+        """Execute process commands or write data to stdin"""
+        if isinstance(msg, dict) and "command" in msg:
+            if msg["command"] == "start":
+                self.start(msg["executable"], msg["args"])
+        elif self.pipe_stdin:
+            if self.process is None:
+                raise CCProcessError("Process is not running.")
+            log.debug(f"write stdin: {msg}")
+            self.process.stdin.write(msg)
+            self.process.stdin.flush()
+        else:
+            raise CCProcessError("Can not send to stdin because pipe is not enabled.")
+
+    def _cleanup(self):
+        """Cleanup threads and process objects"""
+        if self.process is not None:
+            self.process.terminate()
+            self.process.wait()
+        if self.stdout_thread is not None:
+            self.stdout_thread.join()
+            self.stdout_thread = None
+        if self.stderr_thread is not None:
+            self.stderr_thread.join()
+            self.stderr_thread = None
+        if self.queue_in is not None:
+            self.queue_in = None
+
+    def _cc_open(self) -> None:
+        """Implement abstract method"""
+        pass
+
+    def _read_existing(self) -> List[ProcessMessage]:
+        """Read buffered messages that where already received from the process
+
+        :return: Existing messages as tuple: (streamname, data)
+        """
+        messages = self.buffer
+
+        # Get all messages from the process that are available
+        while not self.queue_in.empty():
+            messages.append(self.queue_in.get_nowait())
+        i = 1
+        # Find messages from the same stream(first entry in the tuple) as the first message
+        while (
```

Comment:
Wouldn't it be enough to find the first occurrence of None in the list?
`last_msg_idx = messages.index(None)`

Reply:
No, possible values are: stdout, stderr, None. So if the buffer contains stdout, stdout, stderr, None it should return the first two entries. But there might be a more elegant way of doing this.

Reply:
You could use
```python
initial_stream = messages[0]
itertools.dropwhile(lambda m: m is not None and m.stream == initial_stream, messages)
```

Reply:
I need both parts though. Basically, I want to split the list at the calculated index.

Reply:
Then:
```python
initial_stream = messages[0].stream
consecutive_messages = list(itertools.dropwhile(lambda m: m is not None and m.stream == initial_stream, messages))
self.buffer = messages[len(consecutive_messages):]
```

If you really want to keep it simple another more readable alternative IMO would be:
```python
if messages[0] is PROCESS_EXIT:
    self.buffer = messages[1:]
    return None
    
message_end_idx  = len(messages)   
initial_stream = messages[0].stream

for i, message in enumerate(messages):
    if message is PROCESS_EXIT or message.stream != initial_stream:
        message_end_idx = i
        break
messages, self.buffer = messages[:message_end_idx], messages[message_end_idx:]
```